### PR TITLE
Add copy/paste ring in source/ring.lisp.

### PR DESCRIPTION
**Pull request by noogie13**
Add `clipboard-ring` ring instance for all interactions with the clipboard from Next.

Copy and paste functions that interact with the ring, as well as paste-from-ring.

### CODE REVIEW

File: source/ring.lisp [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+      (ring-insert ring clipboard-content)))
+  (ring-ref ring 0))
+
+(defun ring-completion-fn (ring)
+  (let ((ring-items (ring-recent-list ring)))
+    (lambda (input)
+      (fuzzy-match input ring-items))))
+
+(define-command paste-from-ring ()
+  "Show `interface' clipboard ring and paste selected entry."
```
Comment (Ambrevar) [inline source/ring.lisp]: `*interface*` (the global).

File: source/ring.lisp [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+(defmethod ring-index ((ring ring) index)
+  "Convert to internal ring index, where items are ordered from newest to oldest."
+  (mod (1- (+ (head-index ring)
+              (- (item-count ring) index)))
+       (ring-size ring)))
+
+(defmethod ring-insert ((ring ring) new-item)
+  "Insert item into RING.
+If RING is full, replace the oldest item."
```
Comment (Ambrevar) [inline source/ring.lisp]: Document returned value.

File: source/ring.lisp [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+  "Minus one from index, with wraparound."
+  (let ((new-index (1- index)))
+    (if (= new-index -1)
+        (1- (length (items ring))) new-index)))
+
+(defmethod ring-index ((ring ring) index)
+  "Convert to internal ring index, where items are ordered from newest to oldest."
+  (mod (1- (+ (head-index ring)
+              (- (item-count ring) index)))
+       (ring-size ring)))
```
Comment (Ambrevar) [inline source/ring.lisp]: Double `ring-index` :p  See line 18.

File: source/ring.lisp [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+              (mod (item-count ring)
+                   (ring-size ring)))
+        new-item)
+  (incf (item-count ring))
+  (prog1 (setf (aref (items ring)
+                     (mod (+ (head-index ring) (item-count ring))
+                          (ring-size ring)))
+               new-item)
+    (if (= (item-count ring) (length (items ring)))
+        (setf (head-index ring) (+1-ring-index ring (head-index ring)))
```
Comment (Ambrevar) [inline source/ring.lisp]: You could probably replace +1-ring-index with `(1+ (head-index ring))`.  I haven't tested, please do :)

File: source/ring.lisp [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+  "Add one to index, with wraparound."
+  (let ((new-index (1+ index)))
+    (if (= new-index (length (items ring)))
+        0 new-index)))
+
+(defmethod -1-ring-index ((ring ring) index)
+  "Minus one from index, with wraparound."
+  (let ((new-index (1- index)))
+    (if (= new-index -1)
+        (1- (length (items ring))) new-index)))
```
Comment (Ambrevar) [inline source/ring.lisp]: -1-ring-index is unused and below I suggested a replacement for +1-ring-index, so in the end you can probably remove those 2 functions.

File: source/ring.lisp (hunk 4 above)
Comment (Ambrevar) [reply source/ring.lisp]: Actually `(incf (head-index ring))` should do!

Comment (noogie13) [reply source/ring.lisp]: Wouldn't that overflow past the bounds of the list? Maybe the `mod` calls elsewhere would cover that though?

Once the list is full the head-index should just be rotating around the list, right?

**Comment by Ambrevar** [comment]
I think you are right, so with a `(mod ... (ring-size ring))` you should
be alright.

File: source/ring.lisp [hunk 6]
```diff
@@ -80,7 +58,7 @@ Return most recent entry in RING."
       (fuzzy-match input ring-items))))
 
 (define-command paste-from-ring ()
-  "Show `interface' clipboard ring and paste selected entry."
+  "Show *interface* clipboard ring and paste selected entry."
```
Comment (Ambrevar) [inline source/ring.lisp]: I meant you should keep the quotes in  `*interface*' :)

**Comment by Ambrevar** [comment]
I think we are good to go (beside the last typo).
Would you like to add some tests?

**Comment by noogie13** [comment]
Wrote some basic tests. Should I improve them?

File: source/ring.lisp (hunk 6 above)
Comment (noogie13) [reply source/ring.lisp]: I wasn't sure about that syntax. Oops ~ fixed it.

**Comment by noogie13** [comment]
I tried to rebase around master and somehow, once again, messed it up with a merge... 

I have no idea how / why it merged? I think somewhere I had some weird refs to another branch?

**Comment by noogie13** [comment]
Ok, but I fixed it. :smiley:

**Comment by Ambrevar** [comment]
Still looking great! :)

**Comment by noogie13** [comment]
Anything to add before merge?

**Comment by Ambrevar** [comment]
Outstanding!  Looking forward to the password manager pull request! :)

**Comment by vindarel** [comment]
Well done Solomon! And with tests even :) <3

**Pull request merged**